### ISSUE TITLE
problems about Inference

### ISSUE THREAD

**Issue by coolcoolhua** [zihangdai/CFO#6]
When running generate_score_data.py under Inference folder, it required cand_rel and can_sub attributes. 
> if hasattr(data, 'cand_sub') and hasattr(data, 'cand_rel') and len(data.cand_rel) > 0 and data.relation in data.cand_rel and data.subject in data.cand_sub:

So I found it can generate these two attributes by query_candidates.py in the same folder, but i requires these documents
>  print 'usage: python query_candidate_relation.py num_processes QAData_cpickle_file attention_score_file [[type_dict]]'

Could anyone tell me how to get attention _score_file ?
Or how to get pickle file containing cand_sub and cand_rel attribute?

**Comment by zihangdai** [comment]
Please refer to line 50 - 55 in `data_preprocess.sh` for the QAData_cpickle_file.

**Comment by coolcoolhua** [comment]
Thank you for your reply!
But when I tried 
>python generate_score_data.py ../SimpleQuestions/PreprocessData/QAData.test.pkl

I cannot get the right output file. All file are just 0 kb.
I tried to print data.cand_sub and data.cand_rel, it just telling QAData has not attribute. Did I use the wrong file to run the python file?

**Comment by zihangdai** [comment]
My guess is the generation of `QAData.test.pkl` in the preprocessing step is faulty such that each entry in the unpickled data does not have required properties (i.e., the condition on line 25 of `generate_score_data.py` evaluates to be `False`). 

The reason why the generation of `QAData.test.pkl` is incorrect is most likely due to errors in the earlier steps of data preprocessing. 

To DEBUG: 
(1) Please make sure you have correctly followed the instructions in `Virtuoso.md`. 
(2) Run each code block of `data_preprocess.sh` one by one, and make sure the output is normal.

**Comment by coolcoolhua** [comment]
1. Well, I used the file QAdata.train.pkl and QAdata.test.pkl for training step in your models and it ran successfully and get the proper accuracy. 

I checked the process_rawdata.py, the function query_golden_subs(line 44) contains  value called candi_sub, but the function didn't showed in other place. 
And the pickle data only contains 
> data_list.append((question, sub, rel, obj, length)) , line 185 in process raw data

these five attributes, I can get these five attributes right ,but not cand_sub and cand_rel. Or did I misunderstand this?

2. Actually I check the codes about attribute cand_sub and cand_rel. It shows that in QAdata.py, only the function called add_candidate(line 21) can generate these two attributes cand_sub and cand_rel , but this function only appeared in query_candidates.py(line 106 and 108). 
So I still cannot get this two attribute without getting the file required called attention_score_file(query_candidates.py, line 145). 

Did I missed some steps that I should have get score file but I didn't? Or it is generated by generate_score_data.py but I didn't get it because of the previous fault?

Thanks for your patence!

**Comment by zihangdai** [comment]
I see the problem. I did a significant update to the repo based on my vague memory of the project. Please take a look at the new `README.md` and let me know if there's any issue.

**Comment by coolcoolhua** [comment]
Thanks for your update! But I still got some problems.

In FocusedLabeling/process_inference.lua, it required files like label.test.t7 or label.valid.t7.
But in generate_inference_data.py, it only got 
>with io.open('inference-data/label.{}.txt'.format(args.split), 'w', encoding='utf8') as fo:
        for data in data_list:
            if data.text_attention_indices:
                fo.write(u'%s\t%s\n' % (data.question, ' '.join([str(index) for index in data.text_attention_indices])))

for label.test.txt and label.valid.txt. But I cannot get .t7 file. 
Could you please check this?

**Comment by zihangdai** [comment]
For the code block
```
cd FocusedLabeling

python generate_inference_data.py --split ${SPLIT}

th process_inference.lua
th infer_crf.lua \
    -testData inference-data/label.${SPLIT}.t7 \
    -modelFile "path-to-pretrained-model"
```
The expected behavior will be:
- `python generate_inference_data.py --split ${SPLIT}` will create the file `FocusedLabeling/inference-data/label.${SPLIT}.txt`;
- `th process_inference.lua` will turn the text file `label.${SPLIT}.txt` into `label.${SPLIT}.t7` in torch format (both in the folder `FocusedLabeling/inference-data`);
- ``th infer_crf.lua ...``  will generate the file `label.result.${SPLIT}` in the folder `FocusedLabeling`.

Could you tell me which step works incorrectly?

**Comment by coolcoolhua** [comment]
Well, when I ran process_inference.lua, the first problem is that it will report error saying 
>attempt to index global 'opt' (a nil value)

so maybe it's missing 
>local opt = cmd:parse(arg)

Thus it may need to change the cmd command to run the file?
Because when I ran the file when adding local opt=cmd:parse(arg), it will print  opt.testSplit as something like 
>inference-data/label.nil.txt

Thank you!

**Comment by zihangdai** [comment]
Thanks for the info. I just fixed the bug and pushed it.

**Comment by coolcoolhua** [comment]
Next problem is when I ran query_candidates, it reports assertion error, and I print out length of the two values, they are  (21687, 20620) for test files(QAData.test.pkl and label.result.test ) and (10845, 10339) for valid files(QAData.valid.pkl and label.result.valid).

I have no idea why it is wrong. Can you help to fix it?





Well, finally I cut the dataset to let them have the same amount. Also, you missed a 'Simplequestions/' before PreprocessData in

>python ../query_candidates.py 6 \
       ../../PreprocessData/QAData.${SPLIT}.pkl \
       ../../FocusedLabeling/label.result.${SPLIT} \
       ../../KnowledgeBase/type.top-500.pkl

**Comment by coolcoolhua** [comment]
Thanks for your guidance！ I got the final result!
